### PR TITLE
C++: convert lib/cogip_defs and lib/obstacles into C++ classes

### PR DESCRIPTION
**Pull request by ecourtois**
Convert all structs from lib/cogip_defs and lib/obstacles into C++ classes.
Split files to have only one class by .hpp/.cpp files.

### CODE REVIEW

File: lib/cogip_defs/include/cogip_defs/Coords.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+/// @author      Eric Courtois <eric.courtois@gmail.com>
+
+#pragma once
+
+namespace cogip {
+
+namespace cogip_defs {
+
+class Coords {
+public:
```
Comment (gdoffe) [inline lib/cogip_defs/include/cogip_defs/Coords.hpp:19]: no indent for namespaces ?

File: lib/cogip_defs/Makefile.include [hunk 2]
```diff
@@ -1,2 +1,5 @@
 USEMODULE_INCLUDES_cogip_defs := $(LAST_MAKEFILEDIR)/include
 USEMODULE_INCLUDES += $(USEMODULE_INCLUDES_cogip_defs)
+
```
Comment (gdoffe) [inline lib/cogip_defs/Makefile.include:3]: Should be in a seperated previous commit

File: lib/obstacles/include/obstacles/List.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    ~List();
+
+    uint32_t default_circle_radius() { return default_circle_radius_; };
+    uint32_t default_rectangle_width() { return default_rectangle_width_; };
+    uint32_t min_distance() { return min_distance_; };
+    uint32_t max_distance() { return max_distance_; };
+    void lock() { mutex_.lock(); };
+    void unlock() { mutex_.unlock(); };
+    void print_json(cogip::tracefd::File &out) const;
+    void clear();
```
Comment (gdoffe) [inline lib/obstacles/include/obstacles/List.hpp:45]: Add doxygen please

File: lib/obstacles/include/obstacles/Rectangle.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+namespace cogip {
+
+namespace obstacles {
+
+class Rectangle : public Polygon {
+public:
+    Rectangle(const cogip_defs::Coords &center, double angle,
+              double length_x, double length_y);
+
+    void print_json(cogip::tracefd::File &out) const;
```
Comment (gdoffe) [inline lib/obstacles/include/obstacles/Rectangle.hpp:25]: missing doxygen

File: lib/obstacles/include/obstacles/Polygon.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+class Polygon : public Obstacle, public cogip_defs::Polygon {
+public:
+    Polygon(const std::list<cogip_defs::Coords> *points = nullptr);
+
+    bool is_point_inside(const cogip_defs::Coords &p) const;
+    bool is_segment_crossing(const cogip_defs::Coords &a, const cogip_defs::Coords &b) const;
+    cogip_defs::Coords nearest_point(const cogip_defs::Coords &p) const;
+    void print_json(cogip::tracefd::File &out) const;
+};
```
Comment (gdoffe) [inline lib/obstacles/include/obstacles/Polygon.hpp:31]: missing doxygen

File: lib/obstacles/include/obstacles/Circle.hpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+namespace obstacles {
+
+class Circle : public Obstacle {
+public:
+    Circle(const cogip::cogip_defs::Coords &center, double radius, double angle);
+
+    bool is_point_inside(const cogip_defs::Coords &p) const;
+    bool is_segment_crossing(const cogip_defs::Coords &a, const cogip_defs::Coords &b) const;
+    cogip_defs::Coords nearest_point(const cogip_defs::Coords &p) const;
+    void print_json(cogip::tracefd::File &out) const;
```
Comment (gdoffe) [inline lib/obstacles/include/obstacles/Circle.hpp:27]: missing doxygen

File: lib/cogip_defs/include/cogip_defs/Coords.hpp (hunk 1 above)
Comment (ecourtois) [reply lib/cogip_defs/include/cogip_defs/Coords.hpp:19]: Nope, never ;-)

File: lib/obstacles/include/obstacles/List.hpp (hunk 3 above)
Comment (ecourtois) [reply lib/obstacles/include/obstacles/List.hpp:45]: Done

File: lib/obstacles/include/obstacles/Rectangle.hpp (hunk 4 above)
Comment (ecourtois) [reply lib/obstacles/include/obstacles/Rectangle.hpp:25]: Done

File: lib/obstacles/include/obstacles/Polygon.hpp (hunk 5 above)
Comment (ecourtois) [reply lib/obstacles/include/obstacles/Polygon.hpp:31]: Done

File: lib/obstacles/include/obstacles/Circle.hpp (hunk 6 above)
Comment (ecourtois) [reply lib/obstacles/include/obstacles/Circle.hpp:27]: Done

File: lib/cogip_defs/Makefile.include (hunk 2 above)
Comment (ecourtois) [reply lib/cogip_defs/Makefile.include:3]: Done

File: sys/shell_menu/include/shell_menu/shell_menu.hpp [hunk 7]
```diff
@@ -16,30 +16,29 @@
 #include "shell_menu/Command.hpp"
 #include "shell_menu/Menu.hpp"
 
-/// @def NB_SHELL_COMMANDS
-/// @brief Max shell commands by menu
 #ifndef NB_SHELL_COMMANDS
-#  define NB_SHELL_COMMANDS 20
+#  define NB_SHELL_COMMANDS 20    ///< max shell commands by menu
```
Comment (gdoffe) [inline sys/shell_menu/include/shell_menu/shell_menu.hpp:20]: per menu ?

File: sys/tracefd/tracefd_private.hpp [hunk 8]
```diff
@@ -22,9 +22,10 @@ namespace tracefd {
 
 class File;
 
-/// List of all initialized files
+/// List of all initialized files.
 extern std::vector<File *> all_files;
 
+/// Initialize root directory.
```
Comment (gdoffe) [inline sys/tracefd/tracefd_private.hpp:28]: missing @return

**Review by gdoffe — APPROVED** [review]
LGTM.

Just 2 possibles fixes in doc. I let you check and merge, no need to review again.

Here: https://github.com/cogip/mcu-firmware/pull/46/commits/779e88e4de0248b088d7ac7eb2b5ac819a0cdb9b
"for this module" not needed in commit message as "sys:" already precises the scope.

File: sys/shell_menu/include/shell_menu/shell_menu.hpp (hunk 7 above)
Comment (ecourtois) [reply sys/shell_menu/include/shell_menu/shell_menu.hpp:20]: Done

File: sys/tracefd/tracefd_private.hpp (hunk 8 above)
Comment (ecourtois) [reply sys/tracefd/tracefd_private.hpp:28]: Done

**Pull request merged**